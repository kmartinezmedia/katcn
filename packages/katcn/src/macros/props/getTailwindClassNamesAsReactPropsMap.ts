import { SyntaxKind } from 'ts-morph';
import type { AllStyleProps } from '#types';
import { getTwPrefixAndPropName } from '../getTwPrefixAndPropName';
import { katcnTypesSourceFile } from '../tsMorph';

export type TwAsPropsMap = Record<string, AllStyleProps>;

/**
 * Used in tooling to convert tailwind classnames to react props
 *
 * This lookup map be used in conjuction with other ts-morph tooling
 * to convert `<div className="leading-tight" />` to
 * <Text lineHeight="tight" /> which then renders as <Text className="leading-tight" />
 *
 *
 * Returns:
 *  { leading: {lineHeight: 'tight'} }
 */
export function getTailwindClassNamesAsReactPropsMap() {
  const twToPropsMap: TwAsPropsMap = {};
  const typeRefs =
    katcnTypesSourceFile
      .getInterface('AllStyleProps')
      ?.getDescendantsOfKind(SyntaxKind.TypeReference) ?? [];

  for (const typeRef of typeRefs) {
    const identifier = typeRef.getFirstChildByKind(SyntaxKind.Identifier);
    if (!identifier) continue;
    const symbol = identifier.getSymbol();
    const members = symbol?.getMembers();
    if (!members) continue;
    const signatures = members.flatMap((member) =>
      member
        .getDeclarations()
        .filter((item) => !!item.isKind(SyntaxKind.PropertySignature)),
    );
    for (const signature of signatures) {
      const { twPrefix, name } = getTwPrefixAndPropName(signature);
      const type = signature.getType();

      if (type.isUnion()) {
        for (const unionType of type.getUnionTypes()) {
          if (unionType.isBooleanLiteral()) {
            twToPropsMap[twPrefix] = { [name]: true };
          }
          if (unionType.isStringLiteral()) {
            const propValue = unionType.getLiteralValue();
            if (typeof propValue === 'string') {
              const classname =
                twPrefix === '' ? `${propValue}` : `${twPrefix}-${propValue}`;
              twToPropsMap[classname] = { [name]: propValue };
            }
          }
        }
      }
    }
  }

  return twToPropsMap;
}
