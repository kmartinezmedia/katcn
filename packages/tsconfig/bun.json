{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "Bun",
  "extends": "./node.json",
  "compilerOptions": {
    "types": ["bun-types"]
  }
}
