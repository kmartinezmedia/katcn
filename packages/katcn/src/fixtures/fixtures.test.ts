/// <reference types="bun-types" />

import { describe, expect, it } from 'bun:test';
import { defaultPropsForComponentMap, htmlToComponentMap } from './html';
import {
  modifiers,
  reactModifierPropsToTailwindModifierClassNamesMap,
  tailwindModifierClassNamesToReactPropsMap,
} from './modifiers';
import { styleProps } from './props';

describe('fixtures', () => {
  it('modifiers > modifiers matches snapshot', () => {
    expect(modifiers).toMatchSnapshot();
  });
  it('modifiers > reactModifierPropsToTailwindModifierClassNamesMap', () => {
    expect(reactModifierPropsToTailwindModifierClassNamesMap).toMatchSnapshot();
  });
  it('modifiers > tailwindModifierClassNamesToReactPropsMap', () => {
    expect(tailwindModifierClassNamesToReactPropsMap).toMatchSnapshot();
  });
  it('props > styleProps matches snapshot', () => {
    expect(styleProps).toMatchSnapshot();
  });
  it('html > htmlToComponentMap matches snapshot', () => {
    expect(htmlToComponentMap).toMatchSnapshot();
  });
  it('html > defaultPropsForComponentMap matches snapshot', () => {
    expect(defaultPropsForComponentMap).toMatchSnapshot();
  });
});
