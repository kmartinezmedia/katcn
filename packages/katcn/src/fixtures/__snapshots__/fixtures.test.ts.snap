// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`fixtures modifiers > modifiers matches snapshot 1`] = `
[
  "_after",
  "_dark",
  "_hover",
  "_focus",
  "_focusWithin",
  "_focusVisible",
  "_active",
  "_visited",
  "_target",
  "_first",
  "_last",
  "_only",
  "_odd",
  "_even",
  "_firstOfType",
  "_lastOfType",
  "_onlyOfType",
  "_empty",
  "_disabled",
  "_enabled",
  "_checked",
  "_indeterminate",
  "_default",
  "_required",
  "_valid",
  "_invalid",
  "_inRange",
  "_outOfRange",
  "_placeholderShown",
  "_autofill",
  "_readonly",
  "_file",
  "_selection",
  "_backdrop",
  "_sm",
  "_md",
  "_lg",
  "_xl",
  "_2xl",
  "_motionReduce",
  "_motionSafe",
  "_contrastMore",
  "_contrastLess",
  "_forcedColors",
  "_placeholder",
  "_ariaBusy",
  "_ariaChecked",
  "_ariaHidden",
  "_ariaPressed",
  "_ariaReadonly",
  "_ariaRequired",
  "_ariaSelected",
  "_ariaDisabled",
  "_ariaExpanded",
  "_dataDisabled",
  "_cmdkGroup",
  "_cmdkGroupHeading",
  "_cmdkGroupItems",
  "_cmdkItem",
  "_cmdkItemSvg",
  "_cmdkInput",
  "_cmdkInputSvg",
]
`;

exports[`fixtures modifiers > reactModifierPropsToTailwindModifierClassNamesMap 1`] = `
{
  "_2xl": "2xl:",
  "_active": "active:",
  "_after": "after:",
  "_ariaBusy": "aria-busy:",
  "_ariaChecked": "aria-checked:",
  "_ariaDisabled": "aria-disabled:",
  "_ariaExpanded": "aria-expanded:",
  "_ariaHidden": "aria-hidden:",
  "_ariaPressed": "aria-pressed:",
  "_ariaReadonly": "aria-readonly:",
  "_ariaRequired": "aria-required:",
  "_ariaSelected": "aria-selected:",
  "_autofill": "autofill:",
  "_backdrop": "backdrop:",
  "_checked": "checked:",
  "_cmdkGroup": "cmdkGroup:",
  "_cmdkGroupHeading": "cmdkGroupHeading:",
  "_cmdkGroupItems": "cmdkGroupItems:",
  "_cmdkInput": "cmdkInput:",
  "_cmdkInputSvg": "cmdkInputSvg:",
  "_cmdkItem": "cmdkItem:",
  "_cmdkItemSvg": "cmdkItemSvg:",
  "_contrastLess": "contrast-less:",
  "_contrastMore": "contrast-more:",
  "_dark": "dark:",
  "_dataDisabled": "data-[disabled=true]:",
  "_default": "default:",
  "_disabled": "disabled:",
  "_empty": "empty:",
  "_enabled": "enabled:",
  "_even": "even:",
  "_file": "file:",
  "_first": "first:",
  "_firstOfType": "first-of-type:",
  "_focus": "focus:",
  "_focusVisible": "focus-visible:",
  "_focusWithin": "focus-within:",
  "_forcedColors": "forced-colors:",
  "_hover": "hover:",
  "_inRange": "in-range:",
  "_indeterminate": "indeterminate:",
  "_invalid": "invalid:",
  "_last": "last:",
  "_lastOfType": "last-of-type:",
  "_lg": "lg:",
  "_md": "md:",
  "_motionReduce": "motion-reduce:",
  "_motionSafe": "motion-safe:",
  "_odd": "odd:",
  "_only": "only:",
  "_onlyOfType": "only-of-type:",
  "_outOfRange": "out-of-range:",
  "_placeholder": "placeholder:",
  "_placeholderShown": "placeholder-shown:",
  "_readonly": "read-only:",
  "_required": "required:",
  "_selection": "selection:",
  "_sm": "sm:",
  "_target": "target:",
  "_valid": "valid:",
  "_visited": "visited:",
  "_xl": "xl:",
}
`;

exports[`fixtures modifiers > tailwindModifierClassNamesToReactPropsMap 1`] = `
{
  "2xl": "_2xl",
  "active": "_active",
  "after": "_after",
  'aria-busy": "_ariaBusy",
  'aria-checked": "_ariaChecked",
  'aria-disabled": "_ariaDisabled",
  'aria-expanded": "_ariaExpanded",
  'aria-hidden": "_ariaHidden",
  'aria-pressed": "_ariaPressed",
  'aria-readonly": "_ariaReadonly",
  'aria-required": "_ariaRequired",
  'aria-selected": "_ariaSelected",
  "autofill": "_autofill",
  "backdrop": "_backdrop",
  "checked": "_checked",
  "cmdkGroup": "_cmdkGroup",
  "cmdkGroupHeading": "_cmdkGroupHeading",
  "cmdkGroupItems": "_cmdkGroupItems",
  "cmdkInput": "_cmdkInput",
  "cmdkInputSvg": "_cmdkInputSvg",
  "cmdkItem": "_cmdkItem",
  "cmdkItemSvg": "_cmdkItemSvg",
  'contrast-less": "_contrastLess",
  'contrast-more": "_contrastMore",
  "dark": "_dark",
  'data-[disabled=true]": "_dataDisabled",
  "default": "_default",
  "disabled": "_disabled",
  "empty": "_empty",
  "enabled": "_enabled",
  "even": "_even",
  "file": "_file",
  "first": "_first",
  'first-of-type": "_firstOfType",
  "focus": "_focus",
  'focus-visible": "_focusVisible",
  'focus-within": "_focusWithin",
  'forced-colors": "_forcedColors",
  "hover": "_hover",
  'in-range": "_inRange",
  "indeterminate": "_indeterminate",
  "invalid": "_invalid",
  "last": "_last",
  'last-of-type": "_lastOfType",
  "lg": "_lg",
  "md": "_md",
  'motion-reduce": "_motionReduce",
  'motion-safe": "_motionSafe",
  "odd": "_odd",
  "only": "_only",
  'only-of-type": "_onlyOfType",
  'out-of-range": "_outOfRange",
  "placeholder": "_placeholder",
  'placeholder-shown": "_placeholderShown",
  'read-only": "_readonly",
  "required": "_required",
  "selection": "_selection",
  "sm": "_sm",
  "target": "_target",
  "valid": "_valid",
  "visited": "_visited",
  "xl": "_xl",
}
`;

exports[`fixtures props > styleProps matches snapshot 1`] = `
[
  "aspectRatio",
  "boxSizing",
  "container",
  "columns",
  "display",
  "isolate",
  "objectFit",
  "objectPosition",
  "overflow",
  "overflowX",
  "overflowY",
  "overscrollBehavior",
  "position",
  "top",
  "bottom",
  "left",
  "right",
  "inset",
  "insetX",
  "insetY",
  "visibility",
  "zIndex",
  "scale",
  "scaleX",
  "scaleY",
  "rotate",
  "skewX",
  "skewY",
  "translateX",
  "translateY",
  "transformOrigin",
  "bg",
  "flex",
  "flexDirection",
  "flexGrow",
  "flexShrink",
  "flexWrap",
  "justifyContent",
  "alignItems",
  "alignSelf",
  "alignContent",
  "placeContent",
  "placeItems",
  "placeSelf",
  "shadow",
  "shadowColor",
  "opacity",
  "mixBlendMode",
  "bgBlendMode",
  "rounded",
  "roundedTop",
  "roundedBottom",
  "roundedStart",
  "roundedEnd",
  "roundedTopStart",
  "roundedTopEnd",
  "roundedBottomStart",
  "roundedBottomEnd",
  "borderColor",
  "borderYColor",
  "borderXColor",
  "borderStartColor",
  "borderLeftColor",
  "borderEndColor",
  "borderRightColor",
  "borderTopColor",
  "borderBottomColor",
  "border",
  "borderBottom",
  "borderTop",
  "borderLeft",
  "borderRight",
  "borderStart",
  "borderEnd",
  "borderX",
  "borderY",
  "borderWidth",
  "borderYWidth",
  "borderXWidth",
  "borderStartWidth",
  "borderLeftWidth",
  "borderEndWidth",
  "borderRightWidth",
  "borderTopWidth",
  "borderBottomWidth",
  "divideWidth",
  "divideColor",
  "divideStyle",
  "outlineWidth",
  "outlineColor",
  "outlineStyle",
  "outlineOffset",
  "ringWidth",
  "ringColor",
  "ringOffsetWidth",
  "ringOffsetColor",
  "ringInset",
  "spacing",
  "spacingX",
  "spacingY",
  "spacingRight",
  "spacingEnd",
  "spacingLeft",
  "spacingStart",
  "spacingTop",
  "spacingBottom",
  "offset",
  "offsetY",
  "offsetX",
  "offsetBottom",
  "offsetEnd",
  "offsetRight",
  "offsetStart",
  "offsetLeft",
  "offsetTop",
  "margin",
  "marginY",
  "marginX",
  "marginBottom",
  "marginEnd",
  "marginRight",
  "marginStart",
  "marginLeft",
  "marginTop",
  "gap",
  "gapX",
  "gapY",
  "spaceX",
  "spaceY",
  "textVariant",
  "color",
  "fontFamily",
  "fontSize",
  "fontSmoothing",
  "fontWeight",
  "lineHeight",
  "textAlign",
  "textTransform",
  "lineClamp",
  "textDecoration",
  "textDecorationColor",
  "textDecorationStyle",
  "textDecorationThickness",
  "textUnderlineOffset",
  "textOverflow",
  "truncate",
  "textWrap",
  "indent",
  "verticalAlign",
  "whitespace",
  "break",
  "hyphens",
  "height",
  "minHeight",
  "maxHeight",
  "width",
  "minWidth",
  "maxWidth",
  "size",
  "transition",
  "transitionDuration",
  "transitionTiming",
  "transitionDelay",
  "animation",
  "appearance",
  "cursor",
  "caretColor",
  "pointerEvents",
  "resize",
  "scrollBehavior",
  "scrollSnap",
  "scrollSnapGap",
  "scrollSnapGapTop",
  "scrollSnapGapBottom",
  "scrollSnapGapStart",
  "scrollSnapGapEnd",
  "scrollSnapGapX",
  "scrollSnapGapY",
  "scrollSnapSpacing",
  "scrollSnapSpacingTop",
  "scrollSnapSpacingBottom",
  "scrollSnapSpacingStart",
  "scrollSnapSpacingEnd",
  "scrollSnapSpacingX",
  "scrollSnapSpacingY",
  "scrollSnapAlign",
  "scrollSnapStop",
  "touchAction",
  "userSelect",
  "willChange",
  "tableBorderCollapse",
  "tableBorderSpacing",
  "tableBorderSpacingX",
  "tableBorderSpacingY",
  "tableLayout",
  "tableCaptionSide",
  "fill",
  "strokeColor",
  "strokeWidth",
  "group",
  "className",
  "_after",
  "_dark",
  "_hover",
  "_focus",
  "_focusWithin",
  "_focusVisible",
  "_active",
  "_visited",
  "_target",
  "_first",
  "_last",
  "_only",
  "_odd",
  "_even",
  "_firstOfType",
  "_lastOfType",
  "_onlyOfType",
  "_empty",
  "_disabled",
  "_enabled",
  "_checked",
  "_indeterminate",
  "_default",
  "_required",
  "_valid",
  "_invalid",
  "_inRange",
  "_outOfRange",
  "_placeholderShown",
  "_autofill",
  "_readonly",
  "_file",
  "_selection",
  "_backdrop",
  "_sm",
  "_md",
  "_lg",
  "_xl",
  "_2xl",
  "_motionReduce",
  "_motionSafe",
  "_contrastMore",
  "_contrastLess",
  "_forcedColors",
  "_placeholder",
  "_ariaBusy",
  "_ariaChecked",
  "_ariaHidden",
  "_ariaPressed",
  "_ariaReadonly",
  "_ariaRequired",
  "_ariaSelected",
  "_ariaDisabled",
  "_ariaExpanded",
  "_dataDisabled",
  "_cmdkGroup",
  "_cmdkGroupHeading",
  "_cmdkGroupItems",
  "_cmdkItem",
  "_cmdkItemSvg",
  "_cmdkInput",
  "_cmdkInputSvg",
]
`;

exports[`fixtures html > htmlToComponentMap matches snapshot 1`] = `
{
  "a": "Anchor",
  "article": "Article",
  "aside": "Aside",
  "button": "Pressable",
  "div": "Box",
  "footer": "Footer",
  "h1": "Text",
  "header": "Header",
  "img": "Image",
  "input": "TextInput",
  "li": "ListItem",
  "main": "Main",
  "nav": "Nav",
  "option": "Option",
  "section": "Section",
  "select": "Select",
  "table": "Table",
  "tbody": "TableBody",
  "td": "TableCell",
  "textarea": "TextArea",
  "th": "TableHeadCell",
  "thead": "TableHead",
  "tr": "TableRow",
  "ul": "List",
}
`;

exports[`fixtures html > defaultPropsForComponentMap matches snapshot 1`] = `
{
  "List": {
    "as": "ul",
  },
  "Text": {
    "as": "h1",
  },
}
`;
